analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -29,5 +29,80 @@ public class PublicMethodArgumentsShouldBeCheckedForNull : SymbolicRuleCheck
+            && IsPropertyAccessorAccessibleFromOtherAssemblies(accessor.Modifiers);
+
+        static bool IsAccessibleFromOtherAssemblies(SyntaxNode node) =>
+            node.AncestorsAndSelf().OfType<MemberDeclarationSyntax>().FirstOrDefault() is { } containingMember
+            && node.Ancestors().OfType<BaseTypeDeclarationSyntax>().FirstOrDefault() is { } containingType
+            && IsMemberAccessibleFromOtherAssemblies(containingMember.Modifiers(), containingType)
+            && IsTypeAccessibleFromOtherAssemblies(containingType.Modifiers);
+
+        static bool IsMemberAccessibleFromOtherAssemblies(SyntaxTokenList modifiers, BaseTypeDeclarationSyntax containingType) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || (containingType is InterfaceDeclarationSyntax && HasNoDeclaredAccessabilityModifier(modifiers));
+
+        static bool IsPropertyAccessorAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || HasNoDeclaredAccessabilityModifier(modifiers);
+
+        static bool IsTypeAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers);
+
+        static bool IsPublicOrProtectedOrProtectedInternal(SyntaxTokenList modifiers) =>
+            modifiers.Any(x => x.IsKind(SyntaxKind.PublicKeyword))
+            || (modifiers.Any(x => x.IsKind(SyntaxKind.ProtectedKeyword)) && !modifiers.Any(x => x.IsKind(SyntaxKind.PrivateKeyword)));
+
+        static bool HasNoDeclaredAccessabilityModifier(SyntaxTokenList modifiers) =>
+            !modifiers.Any(x => x.IsKind(SyntaxKind.PrivateKeyword)
+                                || x.IsKind(SyntaxKind.ProtectedKeyword)
+                                || x.IsKind(SyntaxKind.InternalKeyword)
+                                || x.IsKind(SyntaxKind.PublicKeyword));
+
+        static bool MethodDereferencesArguments(BaseMethodDeclarationSyntax method)
+        {
+            var argumentNames = method.ParameterList.Parameters.Select(x => x.Identifier.ValueText).ToArray();
+            var walker = new ArgumentDereferenceWalker(argumentNames);
+            walker.SafeVisit(method);
+            return walker.DereferencesMethodArguments;
+        }
+    }
+
+    private sealed class ArgumentDereferenceWalker : SafeCSharpSyntaxWalker
+    {
+        private readonly string[] argumentNames;
+
+        public bool DereferencesMethodArguments { get; private set; }
+
+        public ArgumentDereferenceWalker(string[] argumentNames) =>
+            this.argumentNames = argumentNames;
+
+        public override void Visit(SyntaxNode node)
+        {
+            if (!DereferencesMethodArguments)
+            {
+                base.Visit(node);
+            }
+        }
+
+        public override void VisitIdentifierName(IdentifierNameSyntax node) =>
```

Comment:
Just a warning: if you somehow have nested `IdentifierNameSyntax` nodes, not calling `base.VisitIdentifierName` would stop the visit in depth, potentially resulting in false negatives.

It happened to me with the `ShouldExecute` of S3655 [here](https://github.com/SonarSource/sonar-dotnet/pull/6918#discussion_r1136797811).

Now, in your scenario I don't see how you can end up in nested identifier names, but it's worth at least a quick investigation.